packages/expo-updates/android/src/main/java/expo/modules/updates/loader/RemoteLoader.java

```diff
@@ -41,6 +41,7 @@
   public interface LoaderCallback {
     void onFailure(Exception e);
     void onSuccess(@Nullable UpdateEntity update);
+    void onAssetLoaded(AssetEntity asset, int assetsLoaded, int totalAssets);
```

Comment:
add a docblock explaining that assetsLoaded includes both assets that were successfully downloaded and that failed to be downloaded. ("loaded" might otherwise imply success)

Reply:
oh, hm, that's a good point. It probably makes sense to actually just call the method only for successful asset downloads.

Reply:
Since the loader first tries downloading all assets before calling `finishWithError`, it would make the progress bar more accurate for onAssetLoaded to be called after failures. It also makes the API more intuitive because assetsLoaded will eventually reach totalAssets. Otherwise, if onAssetLoaded were called only on success, there is no guarantee that assetsLoaded will ever equal totalAssets.

Reply:
I'm not sure I agree with that; if a progress bar reaches 100% (or assetsLoaded == totalAssets) I would expect the operation as a whole to be successful. I think it's unintuitive to expect at that point that there might have been failures.

Maybe a reasonable solution/compromise is to keep the original behavior but add a boolean `success` parameter to the callback, so the caller can decide?

Reply:
It might make more sense to call the callback with successfulAssetCount, failedAssetCount, totalAssetCount.

A progress bar at 75% implies that 25% of the work still needs to be done rather than that 75% of the work is correct. Consider a disk integrity checker -- a progress bar at 75% implies that 75% of the disk (bytes, inodes, files, some unit) has been scanned, not that 75% of the disk is intact.